src/fal/planner/executor.py

```diff
@@ -0,0 +1,101 @@
+from typing import List
+
+from fal.planner.schedule import SUCCESS, NodeQueue
+from fal.planner.tasks import FalHookTask, Node, Task
+from faldbt.project import FalDbt
+
+
+@dataclass
+class FutureGroup:
+    args: argparse.Namespace
+    fal_dbt: FalDbt
+    node: Node
+    executor: Executor
+    futures: list[Future] = field(default_factory=list)
+    status: int = SUCCESS
+
+    def __post_init__(self) -> None:
+        # In the case of us having pre-hooks, this is
+        # where we'll trigger them (and handle the node.task)
+        # below.
+        self._add_tasks(self.node.task)
+
+    def process(self, future: Future) -> None:
+        assert future in self.futures
+
+        self.futures.remove(future)
+        self.status |= future.result()
+        if not isinstance(future.task, FalHookTask):
+            # Once the main task gets completed, we'll populate
+            # the task queue with the post-hooks.
+            self._add_tasks(*self.node.post_hooks)
+
+    def _add_tasks(self, *tasks: Task) -> None:
+        for task in tasks:
+            future = self.executor.submit(
+                task.execute,
+                args=self.args,
+                fal_dbt=self.fal_dbt,
+            )
+            future.task, future.group = task, self
+            self.futures.append(future)
+
+    @property
+    def is_done(self) -> int:
+        return len(self.futures) == 0
+
+
+def parallel_executor(
+    args: argparse.Namespace,
+    fal_dbt: FalDbt,
+    queue: NodeQueue,
+    max_threads: int,
+) -> None:
+    def get_futures(future_groups):
+        return {
+            # Unpack all running futures into a single set
+            # to be consumed by wait().
+            future
+            for future_group in future_groups
+            for future in future_group.futures
+        }
+
+    def create_futures(executor: ThreadPoolExecutor) -> List[FutureGroup]:
+        return [
+            # FutureGroup's are the secondary layer of the executor,
+            # managing the parallelization of tasks.
+            FutureGroup(
+                args,
+                fal_dbt,
+                node=node,
+                executor=executor,
+            )
+            for node in queue.iter_available_nodes()
+        ]
+
+    with ThreadPoolExecutor(max_threads) as executor:
+        future_groups = create_futures(executor)
+        futures = get_futures(future_groups)
+        while futures:
+            # Get the first completed futures, mark them done.
+            completed_futures, _ = wait(futures, return_when=FIRST_COMPLETED)
+            for future in completed_futures:
+                group = future.group
+                group.process(future)
+                if group.is_done:
+                    queue.finish(group.node, status=group.status)
+
+            # And load all the tasks that were blocked by those futures.
+            future_groups.extend(create_futures(executor))
+            futures = get_futures(future_groups)
```

Comment:
If we have more than 1 thread, and we have 2 separate `DBTTask` that are NOT dependent on each other.

Could this mean we are running two dbt tasks at the same time?

Reply:
I did an experiment in this branch and I see that we may be running 2 separate `DBTTask` at the same time. This breaks the dbt thread config dbt will try to use all the threads configured in BOTH runs.

I think a solution is to NOT group nodes at all in the `plan` stage and have a single dbt model per `DBTTask`. Since we handle parallelization ourselves, this will enable 

1. faster running of post-hooks after the corresponding dbt model
2. not break the user expectations of the number of threads being used in the data warehouse

Reply:
The thing about parallelizing dbt runs (happens in both solutions) is that https://docs.getdbt.com/reference/project-configs/target-path gets overwritten for each, meaning we have race conditions for reading the `run_results.json`

Reply:
> Could this mean we are running two dbt tasks at the same time?

Yes!

> This breaks the dbt thread config dbt will try to use all the threads configured in BOTH runs.

That is true and a bad side-effect. 

> I think a solution is to NOT group nodes at all in the plan stage and have a single dbt model per DBTTask. Since we handle parallelization ourselves, this will enable

I haven't thinked it like that, and yes it makes sense. (CC: @turbo1912 @mederka). The grouping makes a decent improvement on short running tasks (since each DBT call have a ~0.5-1 second delay) but I don't see any other way around this problem.

> The thing about parallelizing dbt runs (happens in both solutions) is that https://docs.getdbt.com/reference/project-configs/target-path gets overwritten for each, meaning we have race conditions for reading the run_results.json

I think this is another point which won't go away even if we remove the planner. I already have a hack for this in the main PR that makes it work on tests but we'll need a more permanent solution (see the comment there):
https://github.com/fal-ai/fal/pull/383/files#diff-204bc8d05abda9e5becbbb1d2e7bd91accec607176df85f2286e2d993768412dR112-R118

Reply:
Agreed with everything you said, @isidentical . Thanks for reading! 

I had not seen the hack in the tests, I see. I have asked in the dbt Slack community and someone prompted me to open a GitHub issue asking for a flag to control this. Let's see what the dbt team thinks about this.

Reply:
As we have discussed internally, let's defer the threading issue until the full integration. For the second problem, I'll send a PR that runs DBT through Python so we can get around it.